File: 백준/Silver/15975. 화살표 그리기/화살표 그리기.cc
#include <iostream>
#include <map>
#include <vector>
#include <algorithm>
using namespace std;

int main(int argc, const char * argv[]) {
    cin.tie(NULL);
    ios_base::sync_with_stdio(false);
    
    int N, total=0;
    cin >> N;
    map<int,int> line;
    vector<pair<int,int>> arr;
    vector<vector<int>> pointOf(N+1,vector<int>());
    
    for(int i=0; i<N; i++){
        int coord, color;
        cin >> coord >> color;
        arr.push_back(make_pair(coord, color));
        line[coord] = 1000000000;
    }
    sort(arr.begin(), arr.end());
    
    for(auto point: arr){
        int coord = point.first;
        int color = point.second;
        pointOf[color].push_back(coord);
    }
    
    for(int color=1; color<=N; color++){
        if(pointOf[color].empty()){ continue; }
        if(pointOf[color].size()==1){
            line[pointOf[color][0]] = 0;
        }else{
            for(int i=0; i<pointOf[color].size()-1; i++){
                int curr = pointOf[color][i];
                int next = pointOf[color][i+1];
                int length = next-curr;
                line[curr] = min(line[curr],length);
                line[next] = min(line[next],length);
            }
        }
    }
    
    for(auto point: line){ total += point.second; }
    cout << total << "\n";
    return 0;
}

File: 백준/Silver/15975. 화살표 그리기/화살표 그리기.swift
import Foundation

let N = Int(readLine()!)!
var color = Array(repeating: [Int](), count: N+1)
var arr = [(x:Int, color:Int)]()
var line = Dictionary<Int,Int>()

for _ in 0..<N{
    let info = readLine()!.split(separator: " ").map{Int($0)!}
    let X = info[0]
    let Y = info[1]
    arr.append((X,Y))
    line[X] = 1000000000
}

arr.sort{$0.x < $1.x}

for point in arr{
    color[point.color].append(point.x)
}

for i in 1...N{
    if color[i].count == 1{
        let point = color[i][0]
        line[point]! = 0
    }
    
    for k in stride(from: 0, to: color[i].count-1, by: +1){
        let curr = color[i][k]
        let next = color[i][k+1]
        let length = next-curr
        
        line[curr]! = min(line[curr]!,length)
        line[next]! = min(line[next]!,length)
    }
}
var total = line.values.reduce(0){$0+$1}
print(total)